tool/tsh/common/putty_config_windows.go

```diff
@@ -0,0 +1,365 @@
+	err = tc.WithRootClusterClient(cf.Context, func(clt auth.ClientI) error {
+		cas, err = clt.GetCertAuthorities(cf.Context, types.HostCA, false /* exportSecrets */)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		return nil
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// iterate over all the CAs
+	hostCAPublicKeys := make(map[string][]string)
+	for _, ca := range cas {
+		// if this is either the root or the requested leaf cluster, process it
+		if ca.GetName() == rootClusterName || (wantLeaf && ca.GetName() == cf.LeafClusterName) {
+			// if we've found the requested leaf cluster, mark that we've found it
+			if ca.GetName() == cf.LeafClusterName {
+				leafExists = true
+			}
+			for i, key := range ca.GetTrustedSSHKeyPairs() {
+				log.Debugf("%v CA [%v]: %v", ca.GetName(), i, key)
+				kh, err := sshutils.MarshalKnownHost(sshutils.KnownHost{
+					Hostname:      ca.GetClusterName(),
+					AuthorizedKey: key.PublicKey,
+				})
+				if err != nil {
+					return trace.Wrap(err)
+				}
+				_, _, hostCABytes, _, _, err := ssh.ParseKnownHosts([]byte(kh))
+				if err != nil {
+					return trace.Wrap(err)
+				}
+
+				hostCAPublicKey := strings.TrimPrefix(strings.TrimSpace(string(ssh.MarshalAuthorizedKey(hostCABytes))), constants.SSHRSAType+" ")
+				hostCAPublicKeys[ca.GetName()] = append(hostCAPublicKeys[ca.GetName()], hostCAPublicKey)
+			}
+		}
+	}
+	// update the cert path if a leaf cluster was requested
+	proxyCommandClusterName := rootClusterName
+	if wantLeaf {
+		// if we haven't found the requested leaf cluster, error out
+		if !leafExists {
+			return trace.NotFound("Cannot find registered leaf cluster %q. Use the leaf cluster name as it appears in the output of `tsh clusters`.", cf.LeafClusterName)
+		}
+		proxyCommandClusterName = cf.LeafClusterName
+		certificateFilePath = keypaths.SSHCertPath(keysDir, proxyHost, tc.Config.Username, cf.LeafClusterName)
+	}
+
+	// add all host CA public keys for cluster
+	for cluster, hostCAs := range hostCAPublicKeys {
+		keyName := fmt.Sprintf(`TeleportHostCA-%v`, cluster)
+		for i, publicKey := range hostCAs {
+			// append indices to entries if we have multiple public keys for a CA
+			if len(hostCAs) > 1 {
+				keyName = fmt.Sprintf(`%v-%d`, keyName, i)
+			}
+
+			if err := addHostCAPublicKey(keyName, publicKey, hostname); err != nil {
+				log.Errorf("Failed to add host CA key for %v: %T", cluster, err)
+				return trace.Wrap(err)
+			}
+			log.Debugf("Added/updated host CA key %d for %v", i, cluster)
+		}
+	}
+
+	// format local command string (to run 'tsh proxy ssh')
+	localCommandString := formatLocalCommandString(cf.executablePath, proxyCommandClusterName)
+
+	// add session to registry
+	if err := addPuTTYSession(proxyHost, tc.Config.Host, port, login, ppkFilePath, certificateFilePath, localCommandString, cf.LeafClusterName); err != nil {
+		log.Errorf("Failed to add PuTTY session for %v: %T\n", userHostString, err)
+		return trace.Wrap(err)
+	}
+
```

Comment:
There's a fair amount of logic here that's not covered by tests and will only run on Windows.

Up to you, but I would consider extracting parts of this into an OS-agnostic function that we can write tests against.

Overall, this feels pretty low risk for future breakage, but since it only runs on Windows and we don't do a lot of dev on Windows it's the kind of thing where a regression would be hard to detect. For this reason, the more logic we can cover with automated tests, the better.

I don't remember all the little details of this change, but maybe there's some way you can extract some code that takes as input all of the CAs and produces some output that can be used to write to the registry. It would be hard to test the registry writing, but maybe making some assertions against the intermediate format would be worth while.

Reply:
I've refactored a lot of this in the latest commit and added some tests. It's not perfect, but it's a lot better than it was.